File: app/src/main/java/eu/kanade/tachiyomi/ui/base/presenter/BasePresenter.java
package eu.kanade.tachiyomi.ui.base.presenter;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;

import org.greenrobot.eventbus.EventBus;

import icepick.Icepick;
import nucleus.view.ViewWithPresenter;

public class BasePresenter<V extends ViewWithPresenter> extends RxPresenter<V> {

    private Context context;

    @Override
    protected void onCreate(Bundle savedState) {
        super.onCreate(savedState);
        Icepick.restoreInstanceState(this, savedState);
    }

    @Override
    protected void onSave(@NonNull Bundle state) {
        super.onSave(state);
        Icepick.saveInstanceState(this, state);
    }

    public void registerForEvents() {
        EventBus.getDefault().register(this);
    }

    public void unregisterForEvents() {
        EventBus.getDefault().unregister(this);
    }

    public void setContext(Context applicationContext) {
        context = applicationContext;
    }

    public Context getContext() {
        return context;
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/ui/base/presenter/BasePresenter.kt
package eu.kanade.tachiyomi.ui.base.presenter

import android.os.Bundle
import eu.kanade.core.prefs.PreferenceMutableState
import eu.kanade.tachiyomi.core.preference.Preference
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import nucleus.presenter.RxPresenter

open class BasePresenter<V> : RxPresenter<V>() {

    var presenterScope: CoroutineScope = MainScope()

    override fun onCreate(savedState: Bundle?) {
        try {
            super.onCreate(savedState)
        } catch (e: NullPointerException) {
            // Swallow this error. This should be fixed in the library but since it's not critical
            // (only used by restartables) it should be enough. It saves me a fork.
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        presenterScope.cancel()
    }

    // We're trying to avoid using Rx, so we "undeprecate" this
    @Suppress("DEPRECATION")
    override fun getView(): V? {
        return super.getView()
    }

    fun <T> Preference<T>.asState() = PreferenceMutableState(this, presenterScope)
}
